panic: error relabeling required files

<!-- This form is for bug reports and feature requests ONLY!
Also make sure that you visit our User Guide at https://kubevirt.io/user-guide/
-->

**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind enhancement


**What happened**:

Followed the install instructions under https://kubevirt.io/quickstart_cloud/ on a STACKIT SKE 1.19 Kubernetes. 

Every other pod is running fine, except the virt handler
![image](https://user-images.githubusercontent.com/38325136/120033697-3a8e4c00-bffc-11eb-8391-9ea4df811168.png)

It panics with following error:

```bash
➜  ~ k logs virt-handler-lxxjz -n kubevirt
{"component":"virt-handler","hostname":"shoot--p1t9r9jxg1--kubevirt-worker-tlx5n-z1-5df6d-qmtzl","level":"info","pos":"virt-handler.go:183","timestamp":"2021-05-28T19:31:02.128104Z"}
W0528 19:31:02.131403   63743 client_config.go:614] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
{"component":"virt-handler","level":"info","msg":"set verbosity to 2","pos":"virt-handler.go:427","timestamp":"2021-05-28T19:31:02.153896Z"}
{"component":"virt-handler","level":"info","msg":"Starting collector: node name=shoot--p1t9r9jxg1--kubevirt-worker-tlx5n-z1-5df6d-qmtzl","pos":"prometheus.go:556","timestamp":"2021-05-28T19:31:02.153941Z"}
{"component":"virt-handler","level":"info","msg":"STARTING informer CRDInformer","pos":"virtinformers.go:261","timestamp":"2021-05-28T19:31:02.154020Z"}
{"component":"virt-handler","level":"info","msg":"STARTING informer kubeVirtInformer","pos":"virtinformers.go:261","timestamp":"2021-05-28T19:31:02.154067Z"}
{"component":"virt-handler","level":"info","msg":"STARTING informer vmiInformer-sources","pos":"virtinformers.go:261","timestamp":"2021-05-28T19:31:02.154087Z"}
{"component":"virt-handler","level":"info","msg":"STARTING informer vmiInformer-targets","pos":"virtinformers.go:261","timestamp":"2021-05-28T19:31:02.154114Z"}
{"component":"virt-handler","level":"info","msg":"STARTING informer extensionsKubeVirtCAConfigMapInformer","pos":"virtinformers.go:261","timestamp":"2021-05-28T19:31:02.154133Z"}
{"component":"virt-handler","level":"info","msg":"STARTING informer configMapInformer","pos":"virtinformers.go:261","timestamp":"2021-05-28T19:31:02.154147Z"}
{"component":"virt-handler","level":"info","msg":"metrics: max concurrent requests=3","pos":"virt-handler.go:436","timestamp":"2021-05-28T19:31:02.154221Z"}
{"component":"virt-handler","level":"info","msg":"certificate with common name 'kubevirt.io:system:client:virt-handler' retrieved.","pos":"cert-manager.go:198","timestamp":"2021-05-28T19:31:02.154506Z"}
{"component":"virt-handler","level":"info","msg":"certificate with common name 'kubevirt.io:system:node:virt-handler' retrieved.","pos":"cert-manager.go:198","timestamp":"2021-05-28T19:31:02.154517Z"}
{"component":"virt-handler","level":"info","msg":"SELinux is reported as 'permissive'","pos":"virt-handler.go:335","timestamp":"2021-05-28T19:31:02.160496Z"}
{"component":"virt-handler","level":"info","msg":"Updating cluster config from KubeVirt to resource version '9394'","pos":"config-map.go:516","timestamp":"2021-05-28T19:31:02.161825Z"}
{"component":"virt-handler","level":"info","msg":"set verbosity to 2","pos":"virt-handler.go:427","timestamp":"2021-05-28T19:31:02.161856Z"}
{"component":"virt-handler","level":"warning","msg":"Permissive mode, ignoring 'semodule' failure: out: \"libsemanage.semanage_create_store: Could not access module store at /var/lib/selinux/mcs, or it is not a directory. (Read-only file system).\\nlibsemanage.semanage_direct_connect: could not establish direct connection (Read-only file system).\\n/sbin/semodule:  Could not connect to policy handler\\n\", error: exit status 1","pos":"labels.go:102","timestamp":"2021-05-28T19:31:02.168777Z"}
panic: error relabeling required files: error relabeling file /dev/net/tun with label system_u:object_r:container_file_t:s0. Reason: exit status 1

goroutine 1 [running]:
main.(*virtHandlerApp).Run(0xc000236300)
	cmd/virt-handler/virt-handler.go:346 +0x2c94
main.main()
	cmd/virt-handler/virt-handler.go:546 +0x6e
```

**What you expected to happen**:
That everything is up and running:

By default KubeVirt will deploy 7 pods, 3 services, 1 daemonset, 3 deployment apps, 3 replica sets.

**How to reproduce it (as minimally and precisely as possible)**:


**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`): v0.41.0
- Kubernetes version (use `kubectl version`): 
```bash
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.2", GitCommit:"f5743093fd1c663cb0cbc89748f730662345d44d", GitTreeState:"clean", BuildDate:"2020-09-16T13:41:02Z", GoVersion:"go1.15", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.2", GitCommit:"f5743093fd1c663cb0cbc89748f730662345d44d", GitTreeState:"clean", BuildDate:"2020-09-16T13:32:58Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: STACKIT (Gardener on Opentstack)
- OS (e.g. from /etc/os-release): Flatcar Linux

It is interesting that that we can't change this file:

```
panic: error relabeling required files: error relabeling file /dev/net/tun with label system_u:object_r:container_file_t:s0. Reason: exit status 1
```

This line

```
{"component":"virt-handler","level":"warning","msg":"Permissive mode, ignoring 'semodule' failure: out: \"libsemanage.semanage_create_store: Could not access module store at /var/lib/selinux/mcs, or it is not a directory. (Read-only file system).\\nlibsemanage.semanage_direct_connect: could not establish direct connection (Read-only file system).\\n/sbin/semodule:  Could not connect to policy handler\\n\", error: exit status 1","pos":"labels.go:102","timestamp":"2021-05-28T19:31:02.168777Z"}
```

indicates that selinux may be set up in a non-standard way. Can you influence that somehow on STACKIT?

--- a/cmd/virt-handler/virt-handler.go
+++ b/cmd/virt-handler/virt-handler.go
@@ -26,7 +26,6 @@ import (
 	"io"
 	"net/http"
 	"os"
-	"os/exec"
 	"os/signal"
 	"path/filepath"
 	"runtime"
@@ -380,7 +379,7 @@ func (app *virtHandlerApp) Run() {
 
 		// relabel tun device
 		unprivilegedContainerSELinuxLabel := "system_u:object_r:container_file_t:s0"
-		err = relabelFiles(unprivilegedContainerSELinuxLabel, "/dev/net/tun", "/dev/null")
+		err = selinux.RelabelFiles(unprivilegedContainerSELinuxLabel, se.IsPermissive(), "/dev/net/tun", "/dev/null")
 		if err != nil {
 			panic(fmt.Errorf("error relabeling required files: %v", err))
 		}
@@ -608,18 +607,5 @@ func copy(sourceFile string, targetFile string) error {
 	if err != nil {
 		return fmt.Errorf("failed to make file executable: %v", err)
 	}
-	return nil
-}
-
-func relabelFiles(newLabel string, files ...string) error {
-	relabelArgs := []string{"selinux", "relabel", newLabel}
-	for _, file := range files {
-		cmd := exec.Command("virt-chroot", append(relabelArgs, file)...)
-		err := cmd.Run()
-		if err != nil {
-			return fmt.Errorf("error relabeling file %s with label %s. Reason: %v", file, newLabel, err)
-		}
-	}
-
 	return nil
 }

--- a/pkg/virt-handler/non-root.go
+++ b/pkg/virt-handler/non-root.go
@@ -4,7 +4,6 @@ import (
 	"fmt"
 	"io/ioutil"
 	"os"
-	"os/exec"
 	"path/filepath"
 	"strconv"
 	"strings"
@@ -34,10 +33,10 @@ func changeOwnershipAndRelabel(path string) error {
 		return err
 	}
 
-	_, selinuxEnabled, err := selinux.NewSELinux()
+	seLinux, selinuxEnabled, err := selinux.NewSELinux()
 	if err == nil && selinuxEnabled {
 		unprivilegedContainerSELinuxLabel := "system_u:object_r:container_file_t:s0"
-		err = relabelFiles(unprivilegedContainerSELinuxLabel, filepath.Join(path))
+		err = selinux.RelabelFiles(unprivilegedContainerSELinuxLabel, seLinux.IsPermissive(), filepath.Join(path))
 		if err != nil {
 			return (fmt.Errorf("error relabeling %s: %v", path, err))
 		}
@@ -133,20 +132,6 @@ func (d *VirtualMachineController) nonRootSetup(origVMI, vmi *v1.VirtualMachineI
 	}
 	if err := d.prepareTap(origVMI, res); err != nil {
 		return err
-	}
-	return nil
-}
-
-func relabelFiles(newLabel string, files ...string) error {
-	relabelArgs := []string{"selinux", "relabel", newLabel}
-	for _, file := range files {
-		cmd := exec.Command("virt-chroot", append(relabelArgs, file)...)
-		cmd.Stdout = os.Stdout
-		cmd.Stderr = os.Stderr
-		err := cmd.Run()
-		if err != nil {
-			return fmt.Errorf("error relabeling file %s with label %s. Reason: %v", file, newLabel, err)
-		}
 	}
 	return nil
 }

--- a/pkg/virt-handler/selinux/labels.go
+++ b/pkg/virt-handler/selinux/labels.go
@@ -85,7 +85,7 @@ func (se *SELinuxImpl) semodule(args ...string) (out []byte, err error) {
 	} else if !exists {
 		// on some environments some selinux related binaries are missing, e.g. when the cluster runs in containers (kind).
 		// In such a case, inform the admin, but continue.
-		if se.isPermissive() {
+		if se.IsPermissive() {
 			log.DefaultLogger().Warning("Permissive mode, ignoring missing 'semodule' binary. SELinux policies will not be installed.")
 			return []byte{}, nil
 		}
@@ -98,7 +98,7 @@ func (se *SELinuxImpl) semodule(args ...string) (out []byte, err error) {
 	}
 
 	out, err = se.execFunc(virt_chroot.GetChrootBinaryPath(), argsArray...)
-	if err != nil && se.isPermissive() {
+	if err != nil && se.IsPermissive() {
 		log.DefaultLogger().Warningf("Permissive mode, ignoring 'semodule' failure: out: %q, error: %v", string(out), err)
 		return []byte{}, nil
 	}
@@ -106,7 +106,7 @@ func (se *SELinuxImpl) semodule(args ...string) (out []byte, err error) {
 	return out, err
 }
 
-func (se *SELinuxImpl) isPermissive() bool {
+func (se *SELinuxImpl) IsPermissive() bool {
 	return se.mode == "permissive"
 }
 
@@ -161,4 +161,24 @@ func (se *SELinuxImpl) InstallPolicy(dir string) (err error) {
 type SELinux interface {
 	InstallPolicy(dir string) (err error)
 	Mode() string
+	IsPermissive() bool
+}
+
+func RelabelFiles(newLabel string, continueOnError bool, files ...string) error {
+	relabelArgs := []string{"selinux", "relabel", newLabel}
+	for _, file := range files {
+		cmd := exec.Command("virt-chroot", append(relabelArgs, file)...)
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+		err := cmd.Run()
+		if err != nil {
+			err := fmt.Errorf("error relabeling file %s with label %s. Reason: %v", file, newLabel, err)
+			if !continueOnError {
+				return err
+			} else {
+				log.DefaultLogger().Reason(err).Errorf("Relabeling a file faild, continuing since selinux is permissive.")
+			}
+		}
+	}
+	return nil
 }

